### PR TITLE
[ci-app] Add Retry Mechanism to Requests

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
* Add a single retry if the request to the agent fails due to a socket connection close.

### Motivation
From https://nodejs.org/api/http.html#http_request_reusedsocket:

> When sending request through a keep-alive enabled agent, the underlying socket might be reused. But if server closes connection at unfortunate time, client may run into a 'ECONNRESET' error.

### CODE REVIEW

File: packages/dd-trace/src/exporters/agent/request.js [hunk 1]
```diff
@@ -44,13 +27,46 @@ function request (options, callback) {
...
+  options.headers['Content-Length'] = byteLength(data)
+
+  if (containerId) {
+    options.headers['Datadog-Container-ID'] = containerId
+  }
+  const firstRequest = retriableRequest(options, callback, client, data)
+
+  // The first request will be retried if it fails due to a socket connection close
+  const firstRequestErrorHandler = error => {
+    if (firstRequest.reusedSocket && (error.code === 'ECONNRESET' || error.code === 'EPIPE')) {
```
Comment (juan-fernandez) [inline packages/dd-trace/src/exporters/agent/request.js:57]: I'm occasionally getting `EPIPE` errors as well. From https://nodejs.org/api/errors.html: 
> EPIPE (Broken pipe): A write on a pipe, socket, or FIFO for which there is no process to read the data. Commonly encountered at the net and http layers, indicative that the remote side of the stream being written to has been closed.

File: packages/dd-trace/src/exporters/agent/request.js [hunk 2]
```diff
@@ -44,13 +27,46 @@ function request (options, callback) {
...
+      log.warn('Retrying request due to socket connection error')
```
Comment (rochdev) [inline packages/dd-trace/src/exporters/agent/request.js:58]: I would say to use `debug` level here since it's not actionable.

Comment (juan-fernandez) [reply packages/dd-trace/src/exporters/agent/request.js]: that makes sense. Done

**Comment by juan-fernandez** [comment]
is there any other concern @rochdev ?

**Comment by rochdev** [comment]
Can you add tests for the retry mechanism, and also to make sure that there is no retry loop?

**Comment by juan-fernandez** [comment]
> Can you add tests for the retry mechanism, and also to make sure that there is no retry loop?

added in https://github.com/DataDog/dd-trace-js/pull/1336/commits/0926f9bb01504bbf73571a466a778da9d3b28daa4

**Review by rochdev — APPROVED** [review]

**Pull request merged**